README.md
#**NashVitality iOS Github README File** 
#####**February 18, 2015**

###**Getting Started**

Install Titanium Studio from Appcelerator using the instructions here:
http://docs.appcelerator.com/titanium/3.0/#!/guide/Quick_Start 

Install the following modules using the directions from Appcelerator here:
http://docs.appcelerator.com/titanium/3.0/#!/guide/Using_a_Module

1.  Ti.Styled Label (provides html/css style labeling syntax)
GitHub project:  https://github.com/appcelerator-modules/ti.styledlabel
Direct link to module zip file for installation: https://github.com/appcelerator-modules/ti.styledlabel/releases/download/ios-1.6.1/ti.styledlabel-iphone-1.6.1.zip

2.  benCoding.BasicGeo
GitHub project (forked for 64 bit upgrade): https://github.com/yozef/benCoding.BasicGeo
Direct link to module zip file for installation: https://github.com/yozef/benCoding.BasicGeo/blob/master/IOS/basicGeo/dist/bencoding.basicgeo-iphone-0.89.zip

3.  Titanium Facebook Module, pre-installed.  Activated in modules entry in tiapp.xml setup file.
Documentation: http://docs.appcelerator.com/titanium/3.0/#!/api/Modules.Facebook

As of February 2015, Version 1.5 of the NashVitality App is compatible with iPhones running IOS 7 and 8, and is built using the Titanium 3.4.0.GA SDK and Apple’s XCode 6.

###**Introduction**

The NashVitality Mobile App was initially commissioned by the Nashville/Davidson County Office of the Mayor to provide users with a comprehensive list of outdoor activities and green amenities within Davidson County (http://www.nashville.gov/nashvitality and http://www.nashville.gov/Mayors-Office/Priorities/Health/NashVitality.aspx ).  

The free app’s primary focus is on parks and hiking/biking trails and also includes opportunities for water recreation as well as locations of recycling centers, public art, B-Cycle locations, and information on selected outdoor-related health initiatives, such as the 2014 Walk 100 Miles campaign.  BDY Environmental (http://bdy-inc.com/ ) originally wrote the source code and published the app for iOS and Android using Appcelerator’s Titanium (http://www.appcelerator.com/developers/). 


The app’s name and some graphic elements arose from an earlier NashVitality initiative sponsored by the Metro Nashville Health Department.  Rob Williams Design created icons, most screen layouts, and marketing materials for the app (http://robwilliamsdesign.com/#/nashvitality-app/).


Data sources include hiking trails and park images from Metro Nashville Parks (http://www.nashville.gov/Parks-and-Recreation.aspx), boundaries and basemaps (http://www.nashville.gov/Planning-Department/Mapping-and-GIS.aspx ) from the Metro Nashville Planning Department, Walk Bike Nashville, and field data collected by BDY Environmental.   

The app was released as an open-source project in January 2015.

###**Dependencies/Tools**

In addition to the Titanium Appcelerator development environment and trails/points data, the app is dependent upon several tools and data sources:

1.  Most mapping data were edited using either ESRI’s ArcView (http://www.esri.com/software/arcgis/arcgis-for-desktop ) or Quantum GIS software (http://www2.qgis.org/en/site/).  QGIS was used to export trail and point data as geoJson (http://geojson.org/).  geoJson files were then imported into sqlite (http://www.sqlite.org/) and joined to existing descriptive data tables through views.  For table editing BDY used a combination of BASE (http://menial.co.uk/base/), Microsoft Excel, and text editors.

2.  To display maps, the app utilizes a webview running local html and the Leaflet Javascript mapping library (http://leafletjs.com/).  The local html also uses jQuery 1.8 (http://jquery.com/) for DOM manipulation and the esri-leaflet library (https://github.com/Esri/esri-leaflet) for displaying basemaps published by ArcGIS Server REST endpoints.

3.  The default basemap is provided by the Metro Nashville Planning ArcGIS Server tiled REST service published in the WGS84 Web Mercator Projection (“Google Maps Projection” EPSG 3857, http://spatialreference.org/ref/sr-org/6864/): http://maps.nashville.gov/MetGIS/rest/services/Basemaps/NashvilleBasemapMuted_MSD_WGS84/MapServer

4.  BDY created an offline tiled basemap for use in parks where cell-service/internet connections are spotty.  The offline basemap was generated using MapBox’s TileMill software (https://github.com/mapbox/tilemill).  MBTiles from Tilemill were converted to folders of tiled .png images suitable for use in Leaflet by running the python mbutil.py script (https://github.com/mapbox/mbutil).  Due to the size of the tiled map, images are limited to zoom level 15, approx. 1:18,000 scale.
BDY generated a second tiled image basemap for The Groove, a cycling map created by Walk/Bike Nashville.

5.  Moment.js (http://momentjs.com/), a javascript library for parsing dates is used for sorting newsfeeds and reformatting dates.

6.  The news and events feeds are consumed from the following Metro Nashville website sources:
http://www.nashville.gov/Feeds/NewsEventFeed.aspx?type=NEWS&catid=516
http://www.nashville.gov/Feeds/NewsEventFeed.aspx?type=EVENTS&catid=516


###**Structure**

`App.js` checks for old versions of the installed app sqlite database (e.g., `nashville1.x`) and deletes each previous database if installed.  `App.js` installs version 1.5 of the app database from the Resources folder into the application folder on the mobile device.   `App.js` then creates the Master Window for the app.

`/ui/MasterWindow.js` checks for internet connectivity and basemap availability, and then creates an ios Navigation Window, or navGroup, to keep track of all subsequent windows.  The MasterWindow displays icons for all of the main functionality of the app.  The icons have click events that generate or open each corresponding window.  For subsequent windows that contain maps, the click events first poll the device’s GPS (`/windows/geolocate.js`) to get a location on which to center the map and calculate distances.  The navGroup object is passed to each subsequent window opened. 

#####Project Folder Descriptions

-`Root`: contains `tiapp.xml` project settings file as well as `Resources` folder

-`/Resources`
Contains all project code folders along with `app.js` main/initial controller file, and the app’s `Nashville.db` sqlite file.


-`/Resources/common`
Contains `globals.js` file, mostly used for facebook id

-`/Resources/cs`
Contains deprecated/or Android files

-`/Resources/images`
Contains application image directories

-`/Resources/iphone`
Contains ios app icons and splash screen

-`/Resources/libs`
Contains moment.js javascript library


-`/Resources/mapping`
Contains all html files and libraries used for leaflet based local webview maps.

-`/Resources/pages`
Contains static html pages and css used for webviews (typically about screens, user guides, and more info pages).

-`/Resources/ui`
Contains the `MasterWindow.js` file used to generate the main selection window and navigation controller for the app.

-`/Resources/utility`
Contains deprecated navigation controller code, and code for checking internet connection and whether Metro Nashville ArcGIS Server REST endpoint is available for the basemap.

-`/Resources/windows`
Storage location for most windows and tools in the app.

From the initial Master Window, additional Windows (other than simple information screens with webviews) are opened with the   following pattern:

>`byFoot.js`  --provides a list of maps/activities by foot via a configuration object with image locations, fonts, sql for queries, and map options.  Calls `Generic_Child.js` or a separate window for information display. 

>>`Generic_Child.js` --parses the selected activity and generates a tableview and map/webview for all locations/trails.  Calls the map webview code found in `/Resources/mapping/map.html`. Also uses `/Resources/windows/gc_address_view.js`, `gc_information_view.js`, and `gc_share_view.js` to display additional information about a trail or location.

>OR
  
>>`byFoot_child_....js`--creates a webview or other informational screen.




CHANGELOG.txt
This is Version 1.0 of the NashVitality App for Metro Nashville

Resources/windows/byBike.js
function bkW (gpsInfo, navGroup) {
        var bikeWindow = Titanium.UI.createWindow({
            id: 'byBike',
            title: 'Bike',
            backgroundImage: '/images/backgrounds/background.png',
            barColor: '#D34140',
            translucent: false,
            navTintColor: '#ffffff',
            orientationModes: [Ti.UI.PORTRAIT],
            navBarHidden: false,
            fullscreen: false,
            statusBarStyle: Titanium.UI.iPhone.StatusBar.TRANSLUCENT_BLACK
        });

	

	var data = [
	
		{title:'Explore Near Me',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font: {fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_1.png', leftImage: '/images/tableviews/explorebk.png', hasChild: true,
				headerfont: {fontSize:12,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'select * from vw_explore_bikes',
				theHeaderField: 'Name', theRowHeaderField: 'Type', theRowHeaderTop: 8, theRowNameField: 'Name', 
				//theRowSubHeadingField: 'segment' ,
				latitudeField: 'Lat', longitudeField:'Lon', jsonField: 'parkpoint_json', 
				//addressField: 'Address', cityField: 'City', openField: 'Days', hoursField: 'Hours', 
				//typeField: 'descriptor', summaryField: 'full_name',
				descriptionField: 'site_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg',
				tableIDField: 'table_id', phoneField: 'Phone', emailField: 'email', webField: 'URL', featureType: 'Line',
				parkActivitiesFields : [],				
				 photoField: 'images', theZoomSQL : 'Select * from vw_explore_bikes where table_id = ', backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: true, mainLayer: 'sidewalksbikes',
				mapIconField: 'mapicon', mapTitleField:'Name', mapCategoryField:'Type',  tableIDField: 'table_id', initialfilterDistance: 2}
				//, photoField: 'images'
				},
				
		{title:'Bike Share Program',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_2.png', leftImage: '/images/tableviews/borrow.png', hasChild: true,
				headerfont: {fontSize:16,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'select * from vw_explore_trails where surface="Paved"', 
				orderField: 'th.name', 
				theHeaderField: 'name', theRowHeaderField: 'trailhead_type', theRowHeaderTop: 8, theRowNameField: 'name', 
				theRowSubHeadingField: 'segment' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: 'trailjson', addressField: 'Address', cityField: 'City', openField: 'Days', hoursField: 'hours', typeField: 'descriptor', summaryField: 'full_name',
				descriptionField: 'trail_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', mapIconField: 'mapicon', mapIconURL: '/mapping/redpin.png', 
			 tableIDField: 'trail_id', phoneField: 'phone', emailField: 'email', webField: 'URL', featureType: 'Line',				
				 photoField: 'images', theZoomSQL : 'Select * from vw_explore_trails where trail_id = ', backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: false, mainLayer: 'sidewalksbikes',
				mapIconField: 'mapicon', mapTitleField:'full_name', mapCategoryField:'descriptor',  tableIDField: 'table_id'}
				//, photoField: 'images'
				},
		
				
		{title:'Road Routes',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_3.png', leftImage: '/images/tableviews/road.png', hasChild: true,
				headerfont: {fontSize:16,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140'/*, 
				theSQL : 'select * from vw_explore_trails where surface="Unpaved"', 
				orderField: 'th.name', 
				theHeaderField: 'name', theRowHeaderField: 'trailhead_type', theRowHeaderTop: 8, theRowNameField: 'name', 
				theRowSubHeadingField: 'segment' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: 'trailjson', addressField: 'Address', cityField: 'City', openField: 'Days', hoursField: 'hours', typeField: 'descriptor', summaryField: 'full_name',
				descriptionField: 'trail_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', mapIconField: 'mapicon', mapIconURL: '/mapping/redpin.png', 
				mapTitleField:'full_name', mapCategoryField:'descriptor', popupIconField: 'popupicon', tableIDField: 'trail_id', phoneField: 'phone', emailField: 'email', webField: 'URL', featureType: 'Line',				
				 photoField: 'images', theZoomSQL : 'Select * from vw_explore_trails where trail_id = '
				//, photoField: 'images'*/
				},
				
		{title:'Mtn. Bike Trails',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_4.png', leftImage: '/images/tableviews/mtnbike.png', hasChild: true,
				headerfont: {fontSize:12,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'Select Name || " " || Type || ", " || Segment  as full_name, Type, lat, lon, mtb_id, mtngeojson, site_description, url, images, email, phone, mapicon from mtn_bike_trailheads',
				 orderField: 'name', theHeaderField: 'Type', theRowHeaderField: 'full_name', theRowHeaderTop: 8, //theRowNameField: '', 
				//theRowSubHeadingField: 'segment' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: 'mtngeojson', addressField: '', cityField: '', openField: '', hoursField: '', typeField: '', summaryField: '',
				descriptionField: 'site_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', tableIDField: 'mtb_id', phoneField: 'phone', emailField: 'email', webField: 'url', featureType: 'Point',				
				 photoField: 'images', 
				 theZoomSQL : 'Select Name || " " || Type || ", " || Segment  as full_name, Type, lat, lon, mtb_id, mtngeojson, site_description, url, images, email, phone, mapicon from mtn_bike_trailheads where mtb_id = ',
				  backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: false, mainLayer: 'trails',
				mapIconField: 'mapicon', mapTitleField:'full_name', mapCategoryField:'Type',  tableIDField: 'mtb_id'}
				//, photoField: 'images'
				},
				
		{title:'Metro Greenways',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA',font: {fontSize:18,fontFamily:'Helvetica-Bold' },headerfont: {fontSize:14,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_3.png', leftImage: '/images/tableviews/greenwaysBike.png', hasChild: true, windowBarColor:'#D34140', 
				theSQL : 'select * from vw_explore_trails where category = "Greenway"', orderField: 'name', theHeaderField: 'name', theRowHeaderField: 'name', theRowHeaderTop: 2,theRowNameField: 'segment', 
				theRowSubHeadingField: 'distance' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: 'trailjson', addressField: 'address', cityField: 'city', openField: 'open', hoursField: 'hours', typeField: 'category', summaryField: 'label2',
				descriptionField: 'trail_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg',  tableIDField: 'trail_id', phoneField: 'phone', emailField: 'email', webField: 'webField', photoField: 'photoField', 
				featureType:'Line', theZoomSQL : 'Select * from vw_explore_trails where trail_id = ', backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: false, mainLayer: 'sidewalksbikes',
				mapIconField: 'mapicon_b', mapTitleField:'label', mapCategoryField:'label2',  tableIDField: 'trail_id'}
			},
			
				
		{title:'Music City Bikeway',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_4.png', leftImage: '/images/tableviews/musicbikeway.png', hasChild: true,
				headerfont: {fontSize:12,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'select * from music_city_bikeway', 
				orderField: 'name', theHeaderField: 'name', theRowHeaderField: 'name', theRowHeaderTop: 8, //theRowNameField: 'name', 
				//theRowSubHeadingField: 'segment' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: 'musicbikeway_json', addressField: 'address', cityField: 'city', openField: 'open', hoursField: 'hours', typeField: 'category', summaryField: 'name',
				descriptionField: 'site_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', tableIDField: 'tableid', phoneField: 'phone', emailField: 'email', webField: 'url', featureType: 'Point',				
				 photoField: 'images', theZoomSQL : 'Select * from music_city_bikeway where tableid = ', backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: false, mainLayer: 'sidewalksbikes',
				mapIconField: 'mapicon', mapTitleField:'name', mapCategoryField:'category',  tableIDField: 'tableid'}
				//, photoField: 'images'
				},	
	
		{title:'The Groove',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_4.png', leftImage: '/images/tableviews/groove.png', hasChild: true,
				headerfont: {fontSize:12,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'select * from groove', orderField: 'name', theHeaderField: 'Name', theRowHeaderField: 'Name', theRowHeaderTop: 8, theRowNameField: 'Type2', 
				theRowSubHeadingField: '' ,
				latitudeField: 'lat', longitudeField:'lon', jsonField: '', 
				//addressField: 'Address', cityField: 'City', openField: 'Days', hoursField: 'hours', 
				typeField: 'category', summaryField: 'name',
				descriptionField: 'site_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', mapIconField: 'mapicon', mapIconURL: '/mapping/redpin.png', 
				mapTitleField:'name', mapCategoryField:'category', popupIconField: 'popupicon', tableIDField: 'groove_id', legendPath: '/images/legends/groove_legend.png',
				//phoneField: 'phone', emailField: 'email', webField: 'URL', 
				featureType: 'Line',				
				 photoField: 'images', theZoomSQL : 'Select * from groove where tableid= '
				//, photoField: 'images'
			},
						
		{title:'Bike Parking',backgroundImage:'/images/backgrounds/tableBack.png', height: 64, color:'#0C72BA', font:{fontSize:18,fontFamily:'Helvetica-Bold' },
				selectedBackgroundImage:'/images/tableview/on_4.png', leftImage: '/images/tableviews/bikeParking.png', hasChild: true,
				headerfont: {fontSize:14,fontFamily:'Helvetica-Bold' },subheaderfont: {fontSize:12,fontFamily:'Helvetica-Bold' },windowBarColor:'#D34140', 
				theSQL : 'select * from byBike where Type in ("Bike Share", "Bike Rack")', orderField: 'name', theHeaderField: 'Name', theRowHeaderField: 'Name', theRowHeaderTop: 2, theRowNameField: 'Type2', 
				theRowSubHeadingField: 'artName' ,
				latitudeField: 'Lat', longitudeField:'Lon', jsonField: 'parkpoint_json', 
			
				typeField: 'Type2', summaryField: 'Name',
				descriptionField: 'site_description', defaultPhotoLink: '/images/photos/parks/shelby_btm.jpeg', tableIDField: 'bybike_id', 
				parkActivitiesFields :[],
				featureType: 'Line',				
				 photoField: 'images', theZoomSQL : 'Select Type2, Type, site_description, bybike_id, Name, artName, images, parkpoint_json, Lat, Lon  from byBike where bybike_id = ', backButtonPath: 'images/buttons/back_main.png',
				 mapSettings: {mapIconPath: 'icons/mappins/', polyLineColor: '#ff7800', pointColor: '#ff7800', startLocator: true, mainLayer: 'sidewalksbikes',
				mapIconField: 'mapicon', mapTitleField:'Name', mapCategoryField:'Type2',  tableIDField: 'bybike_id', initialfilterDistance: 2}
				//, photoField: 'images'
				}
				

	];
    
    if (Ti.Platform.osname !== 'android') {
                var leftButton = Ti.UI.createButton({
                    backgroundImage: 'images/buttons/back_main.png',
                    width: 58,
            		height: 20
                });
                leftButton.addEventListener('click', function () {
                   navGroup.closeWindow(bikeWindow);
                });
                bikeWindow.leftNavButton = leftButton;
            }
    
    var tableview = Titanium.UI.createTableView({
        data:data,
        top:0,
        //height:448,
        width:320,
        //color:'#fff',
        //font:{fontSize:18,fontFamily:'Helvetica-Bold',  fontColor:'#0C72BA'},
        backgroundColor:'transparent',
        //backgroundImage:'/images/backgrounds/tableBack.png',
        separatorColor:'#fff'
        });
	bikeWindow.add(tableview);

	if (Ti.Platform.osname == 'ipad')
	{
		tableview.width = 768;
	}


/*tableview.addEventListener('click', function(e)
{
	if (e.rowData.test)
	{
		var win = Titanium.UI.createWindow({
			url:e.rowData.test,
			title:e.rowData.title,barColor:'#000'
		});
		Nash.navGroup.open(win,{animated:true});
	}
	else if (e.rowData.title == 'Explore Near Me') {
		 /*Nash.navGroup.open(Nash.ui.createMetroParksWindow({
                //title: e.rowData.name,
               // uid: e.rowData.uid,
                //name: e.rowData.name
            }), {
                animated: true
            });
		
	}*/
//TABLEVIEW EVENT LISTENER 
//TABLEVIEW EVENT LISTENER 
tableview.addEventListener('click', function(e)
{

	
		
			
		 if (e.rowData.title == "Bike Share Program"){
				var theBikeShareWindow =  require("/windows/byBike_child_share");
				navGroup.openWindow(new theBikeShareWindow(e.rowData, navGroup, gpsInfo));
		}
		
		
		else if (e.rowData.title == "The Groove"){
				var theGrooveWindow =  require("/windows/byBike_child_groove");
				navGroup.openWindow(new theGrooveWindow(e.rowData, navGroup, gpsInfo));

		}
			else if (e.rowData.title == "Road Routes"){
				
				var theRRWindow =  require("/windows/byBike_child_roadroutes");
				navGroup.openWindow(new theRRWindow(e.rowData, navGroup, gpsInfo));
		}
		
		else  {		
			Ti.App.fireEvent('show_indicator', {message: 'Loading Maps and Data . . .'});
			var genericChildWindow = require("/windows/Generic_child");
			navGroup.openWindow(new genericChildWindow(e.rowData, navGroup, gpsInfo));
		
		}

	


	}); 
	
	
    return bikeWindow;
         
}
module.exports = bkW;    
    


